File: app/src/main/java/com/example/garbagesorting/java/ItemsDB.java
package com.example.garbagesorting;

import android.content.Context;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;

public class ItemsDB {
    private static ItemsDB itemsDB;
    private static Context context;
    private final HashMap<String, String> db = new HashMap<String, String>();

    private ItemsDB() {
        if (context == null) {
            throw new IllegalStateException("context must be set first");
        }
        fillItemsDB(context, "items.txt");
    }

    public static void setContext(Context context) {
        ItemsDB.context = context;
    }

    public static ItemsDB get() {
        if (itemsDB == null) itemsDB = new ItemsDB();
        return itemsDB;
    }

    public void addItem(String what, String where) {
        db.put(what, where);
    }

    public String search(String what) {
        return db.getOrDefault(what, "not found");
    }

    private void fillItemsDB(Context context, String fileName) {
        try {
            BufferedReader reader = new BufferedReader(
                    new InputStreamReader(context.getAssets().open(fileName)));
            String line;
            while ((line = reader.readLine()) != null) {
                String[] itemWhatWhere = line.split(", ");
                this.addItem(itemWhatWhere[0], itemWhatWhere[1]);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/example/garbagesorting/ItemsDB.kt
package com.example.garbagesorting

import android.content.Context
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import  java.util.Observable

class ItemsDB private constructor() : Observable() {
    private val db = HashMap<String, String>()

    init {
        fillItemsDB(context!!, "items.txt")
    }

    companion object {
        private var itemsDB: ItemsDB? = null
        private var context: Context? = null

        fun setContext(context: Context) {
            ItemsDB.context = context
        }

        fun get(): ItemsDB {
            return itemsDB ?: ItemsDB().also { itemsDB = it }
        }
    }

    fun listItems(separator: String): List<String> {
        return db.entries.map { "${it.key} $separator ${it.value}" }
    }

    fun addItem(what: String, where: String) {
        db[what] = where
        setChanged()
        notifyObservers()
    }

    fun search(what: String): String {
        return db.getOrDefault(what, "not found")
    }

    private fun fillItemsDB(context: Context, fileName: String) {
        try {
            val reader = BufferedReader(InputStreamReader(context.assets.open(fileName)))
            for (line: String in reader.lines()) {
                val itemWhatWhere = line.split(", ")
                this.addItem(itemWhatWhere[0], itemWhatWhere[1])
            }
            setChanged()
            notifyObservers()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
